Gems/Terrain/Assets/Shaders/Terrain/TerrainMacroTextureClipmapGenerationPass.azsl

```diff
@@ -0,0 +1,127 @@
+    float macroTileSize = TerrainSrg::m_macroMaterialGrid.m_tileSize;
+    float2 macroGridOffset = TerrainSrg::m_macroMaterialGrid.m_offset;
+    uint2 macroGridPosition = (worldPosition - macroGridOffset) / macroTileSize;
+
+    uint macroTileIndex = macroGridResolution.x * macroGridPosition.y + macroGridPosition.x;
+    static const uint NumMacroMaterialsPerTile = 4;
+    macroTileIndex *= NumMacroMaterialsPerTile;
+
+
+    [unroll] for (uint i = 0; i < NumMacroMaterialsPerTile; ++i)
+    {
+        TerrainSrg::MacroMaterialData macroMaterialData = TerrainSrg::m_macroMaterialData[macroTileIndex + i];
+        if ((macroMaterialData.m_flags & 1) == 0)
+        {
+            break; // No more macro materials for this tile
+        }
+
+        if (any(worldPosition < macroMaterialData.m_boundsMin.xy) || any(worldPosition > macroMaterialData.m_boundsMax.xy))
+        {
+            continue; // Macro material exists for this tile but is out of the bounds of this particular position
+        }
+
+        float2 macroUvSize = macroMaterialData.m_boundsMax - macroMaterialData.m_boundsMin;
+        macroUvSize.x = -macroUvSize.x;
+        float2 macroUv = (macroMaterialData.m_boundsMin - worldPosition) / macroUvSize;
+
+        // The macro uv gradient can vary massively over the quad because different pixels may choose different macro materials with different UVs.
+        // To fix, we use the world position scaled by the macro uv scale which should be fairly uniform across macro materials.
+        float2 ddx_macroUv = ddxPosition(pixelPosition, clipmapLevel) / macroUvSize;
+        float2 ddy_macroUv = ddyPosition(pixelPosition, clipmapLevel) / macroUvSize;
+
+        if (macroMaterialData.m_colorMapId != 0xFFFF)
+        {
+            macroColor = TerrainSrg::m_textures[macroMaterialData.m_colorMapId].SampleGrad(TerrainMaterialSrg::m_sampler, macroUv, ddx_macroUv, ddy_macroUv).rgb;
+            macroColor = TransformColor(macroColor, ColorSpaceId::LinearSRGB, ColorSpaceId::ACEScg);
+        }
+
+        if (macroMaterialData.m_normalMapId != 0xFFFF)
+        {
+            bool flipX = macroMaterialData.m_flags & 2;
+            bool flipY = macroMaterialData.m_flags & 4;
+            float factor = macroMaterialData.m_normalFactor;
+
+            float2 sampledValue = TerrainSrg::m_textures[macroMaterialData.m_normalMapId].SampleGrad(TerrainMaterialSrg::m_sampler, macroUv, ddx_macroUv, ddy_macroUv).xy;
+#if AZ_TRAIT_ASTC_COMPRESSION
+            sampledValue =  (sampledValue * 2) - 1;
+#endif
+            if(flipX)
+            {
+                sampledValue.y = -sampledValue.y;
+            }
+            if(flipY)
+            {
+                sampledValue.x = -sampledValue.x;
+            }
+            macroNormal = normalize(GetTangentSpaceNormal_Unnormalized(sampledValue, factor));
+        }
+
+        break;
+    }
+}
+
+[numthreads(32,32,1)]
```

Comment:
This is one of my concern as well. Because updating each clipmap pixel won't be updated every frame.
I'll later convert this to work on a portion of the clipmap, loop a set of texels in one thread.